### PR TITLE
[Bart/Memory] don't create lm_head

### PR DESCRIPTION
**Pull request by sshleifer**
### Summary 

previously we created a `self.lm_head = nn.Linear()` with the exact same weight matrix as the input embeddings, and then skipped `tie_weights`.
This presented 3 problems:
1) 200 MB of extra GPU RAM
2) Can't `tie_weights`
3) Can't `resize_token_embeddings`

This PR alleviates all the concerns by using `lm_logits = F.linear(decoder_outputs[0], self.shared)`. It also adds more aggressive test coverage that `resize_embeddings` is changing the shape of both input and output embeddings.

### Concerns
1) If I recall from an earlier PR, tying the input and output embeddings to a single parameter is unfriendly to torchscript.
However, neither `Bart` before this change nor `T5ForConditionalGeneration`, which uses the `self.lm_head = nn.Linear` technique, pass the common torchscript tests, which suggests that the weight tying in this PR is not removing functionality that existed before it.

2) The failing test here is caused by fact that S3 has `lm_head` in `state_dict`. I will update S3 right before this PR gets merged. 

3) To pass unit tests and use `.generate`, `get_output_embeddings` must return `nn.Linear`. To satisfy this constraint, this PR makes the `nn.Linear` module on the fly when `get_output_embeddings` is called. I think (but am not sure) that this is fine because `resize_token_embeddings` works by resizing the input_embeddings then calling `tie_weights`, and we have stopped skipping `tie_weights`

(Note there is a separate but related issue that `test_common.py::test_resize_embeddings` is shallow, detailed in https://github.com/huggingface/transformers/issues/3378)

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
LGTM! I forgot - why did we skip the weight tying already?

**Comment by sshleifer** [comment]
@LysandreJik because 
```python
        if hasattr(output_embeddings, "out_features") and hasattr(input_embeddings, "num_embeddings"):
            output_embeddings.out_features = input_embeddings.num_embeddings
```
was breaking because bart-large-cnn doesn't have a mask token.
I can investigate more deeply if it's interesting to anyone.

**Review by thomwolf — APPROVED** [review]
Ok for me
